rust/kernel/thread.rs

```diff
@@ -0,0 +1,247 @@
+    /// contexts).
+    pub fn try_new<F>(name: CStr, f: F) -> KernelResult<Self>
+    where
+        F: FnOnce() -> KernelResult<()>,
+        F: Send + 'static,
+    {
+        // Allocate closure here, because this function maybe returns before `rust_thread_func`
+        // (the function that use the closure) get executed.
+        let boxed_fn: Box<dyn FnOnce() -> KernelResult<()> + 'static> = Box::try_new(f)?;
+
+        // Double boxing here because `dyn FnOnce` is a fat pointer, and we can only pass a `usize`
+        // as the `data` for `kthread_create_on_node`.
+        //
+        // We `into_raw` from this side, and will `from_raw` at the other side to transfer the
+        // ownership of the boxed data.
+        let double_box_ptr = Box::into_raw(Box::try_new(boxed_fn)?) as *mut _;
+
+        let result;
+        // SAFETY:
+        //
+        // - `double_box_ptr` is a proper pointer (generated by `Box::into_raw`), and if succeed,
+        // the new thread will get the ownership.
+        //
+        // - `rust_thread_func` is provided by us and handles the dereference of the
+        // `double_box_ptr` (via `Box::from_raw`).
+        unsafe {
+            result = Self::try_new_c_style(name, rust_thread_func, double_box_ptr);
+        }
+
+        if let Err(e) = result {
+            // Creation fails, we need to get back the double boxed closure.
+            //
+            // SAFETY: `double_box_ptr` is a proper pointer generated by a `Box::into_raw()`
+            // from a box created by us. If the thread creation fails, no one will consume that pointer.
+            unsafe {
+                Box::from_raw(double_box_ptr);
+            }
+
+            Err(e)
+        } else {
+            result
+        }
+    }
+
+    /// Wakes up the thread.
+    ///
+    /// Note that a newly created thread (e.g. via [`Thread::try_new`]) will not run until a
+    /// [`Thread::wake_up`] is called.
+    ///
+    /// # Context
+    ///
+    /// This function might sleep, don't call in atomic contexts.
+    pub fn wake_up(&self) {
+        // SAFETY: `task` is a valid pointer to a kernel thread structure, the refcount of which is
+        // increased in `try_new*`, so it won't point to a freed `task_struct`. And it's not
+        // stopped because `stop` will consume the [`Thread`].
+        unsafe {
+            bindings::wake_up_process(self.task);
+        }
+    }
+
+    /// Stops the thread.
+    ///
+    /// - If the thread hasn't been waken up after creation, the thread closure won't be called,
+    /// and will return `EINTR`. Note that a thread may not be waken up even after
+    /// [`Thread::wake_up`] is called.
+    ///
+    /// - Otherwise, wait for the closure to return or the thread `do_exit` itself.
+    ///
+    /// Consume the [`Thread`] so that it's not accessible. Return the result of the thread
+    /// closure (or the exit code in [`KernelResult`] format).
```

Comment:
```suggestion
    /// Consumes the [`Thread`] so that it's not accessible. In case of error, returns
    /// the exit code of the thread.
```

i.e. it sounds otherwise that somehow it can returns any given value as a result when successful.